### PR TITLE
Trajectory samplers now work with mean functions

### LINKED ISSUE #543: Decoupled trajectory samplers for GPflow models do not use mean function
**Describe the bug**
The decoupled trajectory samplers for GP models do not make use of the model's mean function. This is fine for e.g. a constant mean function, but in the case of e.g. a linear mean function the acquisition rule would not select the correct point.

**Expected behaviour**
This should be a fairly minor modification - all that needs to be changed is that the mean function needs to be added to the current trajectories. This should also be checked for other trajectory samplers.

### PR DESCRIPTION
**Pull request by henrymoss**
@sebastianober noticed that I had been stupid and forgotten to take into account mean functions in my decoupled sampling implementation.

This is now fixed and I have added extra tests to make sure this cant happen again!

### CODE REVIEW

**Comment by henrymoss** [comment]
@uri-granta  at the moment the high level FeatureDecompositionTrajectorySampler is defined for any probabilistic model. However, these samplers only really work for very specific models (i.e. FeatureDecompositionInternalDataModel and FeatureDecompositionInducingPointModel). 

Is there a way to specify this requirement for the high level FeatureDecompositionTrajectorySampler rather than at the lower level RandomFourierFeatureTrajectorySampler and DecoupledTrajectorySampler.  At the moment I had to do a slightly weird lazy init for the mean function in the init of FeatureDecompositionTrajectorySampler as general probabilistic models do not have mean functions.

File: tests/unit/models/gpflow/test_sampler.py [hunk 1]
```diff
@@ -292,7 +292,12 @@ def test_rff_trajectory_sampler_returns_trajectory_function_with_correct_shapes(
     sampler = RandomFourierFeatureTrajectorySampler(model, num_features=num_features)
 
     trajectory = sampler.get_trajectory()
-    xs = tf.linspace([-10.0], [10.0], num_evals)  # [N, D]
+    xs = tf.linspace(
```
Comment (uri-granta) [inline tests/unit/models/gpflow/test_sampler.py:295]: (any reason for the reformatting?)

File: tests/unit/models/gpflow/test_models.py [hunk 2]
```diff
@@ -407,6 +407,22 @@ def test_gaussian_process_regression_raises_for_invalid_init() -> None:
...
+
+    m = gpflow.models.GPR((x, y), gpflow.kernels.RBF())
+    model = GaussianProcessRegression(m)
+    assert isinstance(model.get_mean_function(), gpflow.mean_functions.Zero)
+
+    m = gpflow.models.GPR(
+        (x, y), gpflow.kernels.RBF(), mean_function=gpflow.mean_functions.Linear()
+    )
+    model = GaussianProcessRegression(m)
+    assert isinstance(model.get_mean_function(), gpflow.mean_functions.Linear)
```
Comment (uri-granta) [inline tests/unit/models/gpflow/test_models.py:423]: (Why not check that it's the actual mean function in the model, rather than just the same type?)

**Review by uri-granta — APPROVED** [review]
Looks good! (though I wouldn't trust me to catch any other omissions where the mean function needs to be accounted for but isn't)

**Comment by uri-granta** [comment]
> Is there a way to specify this requirement for the high level FeatureDecompositionTrajectorySampler rather than at the lower level RandomFourierFeatureTrajectorySampler and DecoupledTrajectorySampler.

Not sure I fully understand. If FeatureDecompositionTrajectorySampler only supports the same model types as DecoupledTrajectorySampler then why can't you just define it in the same way:

```python
class FeatureDecompositionTrajectorySampler(
    TrajectorySampler[
        Union[
            FeatureDecompositionInducingPointModel,
            FeatureDecompositionInternalDataModel,
        ]],
    ABC,
):
```

**Pull request merged**